### PR TITLE
Cloudwatch: refactor parsing metric data queries

### PR DESCRIPTION
**Pull request by fridgepoet**
**What is this feature?**

This reorganizes the code within `ParseMetricDataQueries` to be roughly:
 1. Decode JSON 
 3. Validate 
 4. Initialize or parse input data (e.g. period in string -> integer) 
 4. Perform migrations

**Why do we need this feature?**

This file and its relatives have the following issues:
- Multiple occurrences of decoding `backend.DataQuery` JSON to our internal typed definition -- this is inefficient and we should be able to decode only once.
- Query validation, default value initialization/parsing, and migrations are mixed together like spaghetti. This could be better organized and separated into these distinct steps.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/cloud-data-sources/issues/62

### CODE REVIEW

File: pkg/tsdb/cloudwatch/models/request_parser.go [hunk 1]
```diff
@@ -66,66 +56,109 @@ func ParseMetricDataQueries(queries []backend.DataQuery, startTime time.Time, en
...
-func migrateLegacyQuery(queries []backend.DataQuery, dynamicLabelsEnabled bool) ([]*backend.DataQuery, error) {
-	migratedQueries := []*backend.DataQuery{}
-	for _, q := range queries {
-		query := q
-		var queryJson *metricsDataQuery
-		err := json.Unmarshal(query.JSON, &queryJson)
-		if err != nil {
-			return nil, err
-		}
+func (q *CloudWatchQuery) migrateStatisticsAndAlias(query metricsDataQuery, dynamicLabelsEnabled bool) {
```
Comment (kevinwcyu) [inline pkg/tsdb/cloudwatch/models/request_parser.go:89]: Is there a reason to call this `migrateStatisticsAndAlias` instead of keeping it as `migrateLegacyQuery`? In the future I'd imagine we'd add additional migrations here as well.

File: pkg/tsdb/cloudwatch/models/request_parser.go [hunk 2]
```diff
@@ -198,68 +215,16 @@ func parseRequestQuery(dataQuery metricsDataQuery, refId string, startTime time.
...
-				return nil, fmt.Errorf("failed to parse period as integer: %v", err)
-			}
-		} else {
-			d, err := time.ParseDuration(p)
+		period, err = strconv.Atoi(periodString)
+		if err != nil {
+			d, err := time.ParseDuration(periodString)
 			if err != nil {
-				return nil, fmt.Errorf("failed to parse period as duration: %v", err)
+				return 0, fmt.Errorf("failed to parse period as duration: %v", err)
```
Comment (kevinwcyu) [inline pkg/tsdb/cloudwatch/models/request_parser.go:222]: Most Go code I've seen usually returns `nil` if there is an error instead of a value. Is there a reason for returning `0` here instead of `nil`?

**Review by kevinwcyu — APPROVED** [review]

File: pkg/tsdb/cloudwatch/models/request_parser.go (hunk 1 above)
Comment (sunker) [inline pkg/tsdb/cloudwatch/models/request_parser.go:89]: Shouldn't methods be in the same file as the struct definition?

**Review by sunker — APPROVED** [review]
LGTM

File: pkg/tsdb/cloudwatch/models/request_parser.go (hunk 2 above)
Comment (fridgepoet) [reply pkg/tsdb/cloudwatch/models/request_parser.go:222]: Yup, I can't actually return `nil` here because of the `int` type. 
The convention is to return the "zero value" alongside an error. Often the zero values are nil, so that's why you've seen that often.

Here's an article that has types and their zero values https://nilsmagnus.github.io/post/nillability-in-go/
TLDR: This is what I remember off the top of my head and then when needed I look up others: Basic types are not nullable. Structs aren't either. Slices and maps are. (Maps are a common "gotcha" -- they need to be initialized, for example ` var fluffy = make(map[string]string)` or `var fluffy = map[string]string{}` before you start inserting into them. Often we forget to do this.)

File: pkg/tsdb/cloudwatch/models/request_parser.go (hunk 1 above)
Comment (fridgepoet) [reply pkg/tsdb/cloudwatch/models/request_parser.go:89]: No reason in particular. I had the outer function `migrateLegacyQuery` at some point and put it back under a different name. I've renamed it now.

Comment (sunker) [reply pkg/tsdb/cloudwatch/models/request_parser.go]: Making the migrations a method is fine to me now that we have a separate package for models, and these migrations methods are not exported. 

Have looked at your commit and it looks good!

**Comment by kylebrandt** [comment]
This seems to have introduced a test that does not consistently pass (non-deterministic) 

`/usr/local/go/bin/go test -timeout 120s -tags integration -run ^Test_ParseMetricDataQueries_migrate_alias_to_label$ github.com/grafana/grafana/pkg/tsdb/cloudwatch/models -v -count=100` 

Fails starting in 96beb3a3f1
